stdlib/open-uri.rb

```diff
@@ -0,0 +1,515 @@
+#     f.each_line {|line| p line}
+#   }
+#
+# The opened file has several getter methods for its meta-information, as
+# follows, since it is extended by OpenURI::Meta.
+#
+#   open("http://www.ruby-lang.org/en") {|f|
+#     f.each_line {|line| p line}
+#     p f.base_uri         # <URI::HTTP:0x40e6ef2 URL:http://www.ruby-lang.org/en/>
+#     p f.content_type     # "text/html"
+#     p f.charset          # "iso-8859-1"
+#     p f.content_encoding # []
+#     p f.last_modified    # Thu Dec 05 02:45:02 UTC 2002
+#   }
+#
+# Additional header fields can be specified by an optional hash argument.
+#
+#   open("http://www.ruby-lang.org/en/",
+#     "User-Agent" => "Ruby/#{RUBY_VERSION}",
+#     "From" => "foo@bar.invalid",
+#     "Referer" => "http://www.ruby-lang.org/") {|f|
+#     # ...
+#   }
+#
+# The environment variables such as http_proxy, https_proxy and ftp_proxy
+# are in effect by default. Here we disable proxy:
+#
+#   open("http://www.ruby-lang.org/en/", :proxy => nil) {|f|
+#     # ...
+#   }
+#
+# See OpenURI::OpenRead.open and Kernel#open for more on available options.
+#
+# URI objects can be opened in a similar way.
+#
+#   uri = URI.parse("http://www.ruby-lang.org/en/")
+#   uri.open {|f|
+#     # ...
+#   }
+#
+# URI objects can be read directly. The returned string is also extended by
+# OpenURI::Meta.
+#
+#   str = uri.read
+#   p str.base_uri
+#
+# Author:: Tanaka Akira <akr@m17n.org>
+
+module OpenURI
+  Options = {
+    :proxy => true,
+    :proxy_http_basic_authentication => true,
+    :progress_proc => true,
+    :content_length_proc => true,
+    :http_basic_authentication => true,
+    :read_timeout => true,
+    :open_timeout => true,
+    :ssl_ca_cert => nil,
+    :ssl_verify_mode => nil,
+    :ftp_active_mode => false,
+    :redirect => true,
+    :encoding => nil,
+  }
+
+  def OpenURI.check_options(options) # :nodoc:
+    options.each {|k, v|
+      next unless Symbol === k
+      unless Options.include? k
+        raise ArgumentError, "unrecognized option: #{k}"
+      end
+    }
+  end
+
+  def OpenURI.scan_open_optional_arguments(*rest) # :nodoc:
+    if !rest.empty? && (String === rest.first || Integer === rest.first)
+      mode = rest.shift
+      if !rest.empty? && Integer === rest.first
+        perm = rest.shift
+      end
+    end
+    return mode, perm, rest
+  end
+
+  def self.open_uri(name, *rest) # :nodoc:
+    io = open_loop(name, {})
+    io.rewind
+    if block_given?
+      begin
+        yield io
+      ensure
+        if io.respond_to? :close!
+          io.close! # Tempfile
+        else
+          io.close if !io.closed?
+        end
+      end
+    else
+      io
+    end
+  end
+
+  def self.open_loop(uri, options) # :nodoc:
+    req = request(uri)
+    # FIXME Handle exception and status code ?
+    data = `req.responseText`
+    buf = Buffer.new
+    buf << data
+    io = buf.io
+    # FIXME Generate a URI object from the uri String
+    #io.base_uri = uri
+    status = `req.status`
```

Comment:
This part is ugly, I want to assign multiples attributes with the request object... do you have a better idea/syntax ?

Reply:
A simply mass-assignment should work, but I'm not sure that it's a good idea
``` ruby
opal-repl
>> a, b = `[1, 2]`
=> [1, 2]
>> a
=> 1
>> b
=> 2
```

Reply:
Ok I will stick with this syntax, now that we have tests we can improve the code